File: src/main/java/org/tasks/intents/TaskIntents.java
package org.tasks.intents;

import android.content.Context;
import android.content.Intent;
import android.support.v4.app.TaskStackBuilder;

import com.google.common.base.Strings;
import com.todoroo.astrid.activity.TaskListActivity;
import com.todoroo.astrid.api.Filter;

public class TaskIntents {

    public static TaskStackBuilder getEditTaskStack(Context context, final Filter filter, final long taskId) {
        return TaskStackBuilder.create(context).addNextIntent(getEditTaskIntent(context, filter, taskId));
    }

    public static Intent getNewTaskIntent(Context context, String filterId) {
        return getEditTaskIntent(context, filterId, 0L);
    }

    public static Intent getEditTaskIntent(Context context, String filterId, long taskId) {
        Intent taskListIntent = getTaskListByIdIntent(context, filterId);
        taskListIntent.putExtra(TaskListActivity.OPEN_TASK, taskId);
        return taskListIntent;
    }

    private static Intent getEditTaskIntent(Context context, final Filter filter, final long taskId) {
        Intent taskListIntent = getTaskListIntent(context, filter);
        taskListIntent.putExtra(TaskListActivity.OPEN_TASK, taskId);
        return taskListIntent;
    }

    public static Intent getTaskListIntent(Context context, final Filter filter) {
        Intent intent = new Intent(context, TaskListActivity.class);
        if (filter != null) {
            intent.putExtra(TaskListActivity.OPEN_FILTER, filter);
        }
        return intent;
    }

    public static Intent getTaskListByIdIntent(Context context, final String filterId) {
        Intent intent = new Intent(context, TaskListActivity.class);
        if (!Strings.isNullOrEmpty(filterId)) {
            intent.putExtra(TaskListActivity.LOAD_FILTER, filterId);
        }
        return intent;
    }
}

File: app/src/main/java/org/tasks/intents/TaskIntents.kt
package org.tasks.intents

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import com.todoroo.astrid.activity.MainActivity
import org.tasks.data.entity.Task
import org.tasks.filters.Filter

object TaskIntents {
    const val FLAGS: Int = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP

    fun getNewTaskIntent(
        context: Context,
        filter: Filter?,
        createSource: String
    ): Intent {
        val intent = Intent(context, MainActivity::class.java)
        if (filter != null) {
            intent.putExtra(MainActivity.OPEN_FILTER, filter)
        }
        intent.putExtra(MainActivity.CREATE_TASK, 0L)
        intent.putExtra(MainActivity.CREATE_SOURCE, createSource)
        intent.putExtra(MainActivity.REMOVE_TASK, true)
        return intent
    }

    fun getEditTaskIntent(context: Context, filter: Filter?, task: Task?): Intent {
        val intent = Intent(context, MainActivity::class.java)
        if (filter != null) {
            intent.putExtra(MainActivity.OPEN_FILTER, filter)
        }
        intent.putExtra(MainActivity.OPEN_TASK, task)
        intent.putExtra(MainActivity.REMOVE_TASK, true)
        return intent
    }

    fun getTaskListIntent(context: Context, filter: Filter?): Intent {
        val intent = Intent(context, MainActivity::class.java)
        intent.setFlags(FLAGS)
        if (filter != null) {
            intent.putExtra(MainActivity.OPEN_FILTER, filter)
        }
        return intent
    }

    fun getTaskListByIdIntent(context: Context, filterId: String?): Intent {
        val intent = Intent(Intent.ACTION_MAIN)
        intent.setComponent(ComponentName(context, MainActivity::class.java))
        intent.putExtra(MainActivity.LOAD_FILTER, filterId)
        return intent
    }
}
